pos-v2/contracts/Protocol.sol

```diff
@@ -0,0 +1,51 @@
+pragma solidity 0.5.16;
+
+contract Protocol {
+
+    event ProtocolVersionChanged(string deploymentSubset, uint256 protocolVersion, uint256 asOfBlock);
+
+    struct DeploymentSubset {
+        uint version;
+        uint asOfBlock;
+    }
+
+    mapping (string => DeploymentSubset) deploymentSubsets;
+    address governor;
+
+    modifier onlyGovernor() {
+        require(msg.sender == governor, "caller is not the protocol contract governor");
+
+        _;
+    }
+
+    constructor(address _governor) public {
+        governor = _governor;
+    }
+
+    function deploymentSubsetExists(string calldata deploymentSubset) external view returns (bool) {
+        return deploymentSubsets[deploymentSubset].version > 0;
+    }
+
+    function setProtocolVersion(string calldata deploymentSubset, uint256 protocolVersion, uint256 asOfBlock) external onlyGovernor {
+        if (deploymentSubsets[deploymentSubset].version == 0) {
+            require(asOfBlock == 0, "initial protocol version must be from block 0");
+        } else {
+            require(asOfBlock > block.number, "protocol update can only take place in the future");
+            require(asOfBlock > deploymentSubsets[deploymentSubset].asOfBlock, "protocol upgrade can only take place after the previous protocol update");
+            require(protocolVersion > deploymentSubsets[deploymentSubset].version, "protocol downgrade is not supported"); // TODO should we enforce it?
+            // TODO check that is it scheduled not to close to the present?
```

Comment:
I would allow setting `version` and `asOfBlock` to a lower value as in these cases:
1. `asOfBlock` and `deploymentSubsets[deploymentSubset].asOfBlock` are both in the future
2. the effective `protocolVersion` is lower than `protocolVersion` where effective `protocolVersion` is any protocol version which was scheduled to apply before `block.number`

example - if we are running on version `1` and  on block `80` we run `setProtocolVersion("z", 3,120) ` (schedule upgrade to protocol version `3` to commence in block `120`), and then in block `90` I send  `setProtocolVersion("z", 2,110) ` the call should succeed overriding the next upgrade version to 2 in block 110. 

This means that any transaction accepted in this contract must first commit upgrade requests to the "current version".

@OdedWx   what do you think?
 ( - we should review the interface of this contract regardless, asking here about the override logic for upgrade requests)

Reply:
@OdedWx  - agrees